classes/solid/terrain/terrain_border.gd

```diff
@@ -0,0 +1,408 @@
+	
+	# Circular edges, this doesn't work rn, I'll work on it later
+	
+#    #for if e_angle > s_angle
+#    while angle < e_angle:
+#        angle += delta
+#        var vert = circumcenter + Vector2(cos(angle), sin(angle)) * distance
+#        verts.append(vert)
+#    #for if e_angle < s_angle
+#    angle = s_angle
+#    while angle > e_angle:
+#        angle += delta
+#        var vert = circumcenter + Vector2(cos(angle), sin(angle)) * distance
+#        verts.append(vert)
+	
+	# Add the endpoint.
+	verts.append(end)
+	# Add the circumcenter (to make a triangle?).
+	verts.append(circumcenter)
+	
+	# Add the triangle we just created to the areas.
+	areas.append({
+		verts = verts,
+		normal = start.direction_to(end).rotated(-PI / 2),
+		clock_dir = sign(delta),
+		type = "trio"
+	})
+
+
+func generate_polygons(lines: Array, texture: Texture, z_index: int):
+	var p_len = lines.size()
+	for ind in range(p_len - 1):
+		# First create quads from each line segment.
+		var verts = _generate_quad(lines, ind)
+		
+		# Find the normal of this line segment too.
+		# TODO: This is also done in _generate_quad. If this loop ever needs
+		# optimizing, here's some work to deduplicate.
+		var normal := _normal_of_segment(lines[ind], lines[(ind + 1) % p_len])
+		
+		# Create a child polygon node.
+		add_child(_create_polygon(verts, normal, z_index, texture, 0))
+
+
+func _generate_quad(chain: Array, start_idx: int, thickness: int = QUAD_RADIUS):
+	# Find the normal of this line segment.
+	var current: Vector2 = chain[start_idx]
+	var next: Vector2 = chain[(start_idx + 1) % chain.size()]
+	var normal := _normal_of_segment(current, next)
+	
+	# The direction to move verts in order to thicken the line.
+	var thicken_vec := normal * thickness
+	
+	# From the line segment, generate a quad with thickness.
+	# The quad's vertices run clockwise around its perimeter.
+	return [
+		# Top edge
+		current + thicken_vec,
+		next + thicken_vec,
+		# Bottom edge
+		next - thicken_vec,
+		current - thicken_vec
+	]
+
+
+func _create_polygon(
+	verts: Array, normal: Vector2,
+	z_order: int,
+	texture: Texture, origin_idx: int
+) -> Polygon2D:
+	# Create a polygon node from the terrain's data....
+	var poly2d = Polygon2D.new()
+	poly2d.texture = texture
+	poly2d.color = root.tint_color if root.tint else Color(1, 1, 1)
+	# ...and the area's data.
+	poly2d.polygon = verts
+	# This one will be important later.
+	poly2d.texture_rotation = -normal.angle() - PI / 2
+
+	# Variables needed to set texture offset.
+	# Rotated versions of the X and Y axes.
+	var unit = Vector2(sin(poly2d.texture_rotation), cos(poly2d.texture_rotation))
+	# Position of the area's origin (chosen from one of the verts).
+	var pos = verts[origin_idx]
+	# Texture offset. This is never set, so this may not be needed?
+	var tex_offset = Vector2(0, 0)
+	
+	# Scale the texture, not sure if this is still needed, it appears not so it is commented for now
+	#if area.type == "trio":
```

Comment:
I've never gotten the scaling to work properly, you can keep it commented if you wish, or delete it.

I'm not even entirely sure if scaling is needed, I think it can look slightly better if it managed to scale well. Though the current result works for most of the time.